internal/http_client.go

```diff
@@ -21,44 +21,123 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"math"
 	"net/http"
+	"strconv"
+	"time"
+
+	"google.golang.org/api/option"
+	"google.golang.org/api/transport"
 )
 
 // HTTPClient is a convenient API to make HTTP calls.
 //
-// This API handles some of the repetitive tasks such as entity serialization and deserialization
-// involved in making HTTP calls. It provides a convenient mechanism to set headers and query
+// This API handles the repetitive tasks such as entity serialization and deserialization
+// when making HTTP calls. It provides a convenient mechanism to set headers and query
 // parameters on outgoing requests, while enforcing that an explicit context is used per request.
-// Responses returned by HTTPClient can be easily parsed as JSON, and provide a simple mechanism to
-// extract error details.
+// Responses returned by HTTPClient can be easily unmarshalled as JSON.
+//
+// HTTPClient also handles automatically retrying failed HTTP requests.
 type HTTPClient struct {
-	Client    *http.Client
-	ErrParser ErrorParser
+	Client      *http.Client
+	RetryConfig *RetryConfig
+	ErrParser   ErrorParser
 }
 
-// Do executes the given Request, and returns a Response.
-func (c *HTTPClient) Do(ctx context.Context, r *Request) (*Response, error) {
-	req, err := r.buildHTTPRequest()
+// NewHTTPClient creates a new HTTPClient using the provided client options and the default
+// RetryConfig.
+//
+// The default RetryConfig retries requests on all low-level network errors as well as on HTTP
+// InternalServerError (500) and ServiceUnavailable (503) errors. Repeatedly failing requests are
+// retried up to 4 times with exponential backoff. Retry delay is never longer than 2 minutes.
+func NewHTTPClient(ctx context.Context, opts ...option.ClientOption) (*HTTPClient, string, error) {
+	hc, endpoint, err := transport.NewHTTPClient(ctx, opts...)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
+	twoMinutes := time.Duration(2) * time.Minute
+	client := &HTTPClient{
+		Client: hc,
+		RetryConfig: &RetryConfig{
+			MaxRetries: 4,
+			CheckForRetry: retryNetworkAndHTTPErrors(
+				http.StatusInternalServerError,
+				http.StatusServiceUnavailable,
+			),
+			ExpBackoffFactor: 0.5,
+			MaxDelay:         &twoMinutes,
+		},
```

Comment:
Should the RetryConfig options be specified or allowed to be overridden by option.ClientOption?

Reply:
`ClientOption` interface can only be meaningfully implemented in the Google API client code (it's one of their interfaces).  For now I think it's sufficient that we can do this:

```
client, err := internal.NewHTTPClient(ctx, opts...)
client.RetryConfig = buildCustomConfig()
```